proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,590 @@
+```
+
+All values can be passed to functions with extension parameters, or we can omit extension parameters and let the compiler resolve the suitable extensions for them. This makes the approach really similar to the way default arguments work in Kotlin.
+
+```kotlin
+fetchById<User>(11829) // compiles since we got evidence of a `Repository<User>` in scope.
+fetchById<User>(11829, UserRepository()) // you can provide it manually.
+```
+
+When `with` is used in class constructors, it is important to **add val to extension class fields** to make sure they are accessible in the scope of the class. Here, the `with` keyword adds the value to the scope of every method in the class. To showcase this, let's say we have a `Validator<A>`, like:
+
+```kotlin
+interface Validator<A> {
+  fun A.isValid(): Boolean
+}
+```
+
+In this scenario, the following classes would be equivalent:
+
+```kotlin
+data class Group<A>(val values: List<A>)
+
+// Kotlin + KEEP-87
+extension class ValidatedRepository<A>(with val V: Validator<A>) : Repository<A> {
+
+    val storedUsers: MutableMap<Int, A> = mutableMapOf() // e.g. users stored in a DB
+
+    override fun loadAll(): Map<Int, A> {
+        return storedUsers.filter { it.value.isValid() }
+    }
+
+    override fun loadById(id: Int): A? {
+        return storedUsers[id]?.let { if (it.isValid()) it else null }
+    }
+
+    override fun A.save() {
+        storedUsers[generateKey(this)] = this
+    }
+}
+
+// Regular Kotlin
+class ValidatedRepository<A>(val V: Validator<A>) : Repository<A> {
+
+    val storedUsers: MutableMap<Int, A> = mutableMapOf() // e.g. users stored in a DB
+
+    override fun loadAll(): Map<Int, A> {
+        with (V) {
+            return storedUsers.filter { it.value.isValid() }
+        }
+    }
+
+    override fun loadById(id: Int): A? {
+        with (V) {
+            return storedUsers[id]?.let { if (it.isValid()) it else null }
+        }
+    }
+
+    override fun A.save() {
+        storedUsers[generateKey(this)] = this
+    }
+}
+```
+
+As you can see in the first example, `A.isValid()` becomes available inside the method’s scope automatically. The equivalent version of doing this without KEEP-87 is to manually add `with (V)` inside each method, as you can see in the second example.  
+
+## Composition and chain of evidences
+
+Constraint interface declarations and extension evidences can encode further constraints on their type parameters so that they can be composed nicely:
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+import com.domain.Group
+
+extension class GroupRepository<A>(with val repoA: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): Map<Int, Group<A>> {
+    return repoA.loadAll().mapValues { Group(it.value) }
+  }
+
+  override fun loadById(id: Int): Group<A>? {
+    return repoA.loadById(id)?.let { Group(it) }
+  }
+
+  override fun Group<A>.save() {
+    this.items.map { repoA.run { it.save() } }
```

Comment:
according to `data class Group<A>(val values: List<A>)`, shouldn't it be `this.values...`